src/server/utils/logging/otel/logger.js

```diff
@@ -0,0 +1,113 @@
+ * permissions and limitations under the License.
+ */
+
+import os from 'node:os';
+import {
+  LoggerProvider,
+  ConsoleLogRecordExporter,
+  SimpleLogRecordProcessor,
+  BatchLogRecordProcessor,
+} from '@opentelemetry/sdk-logs';
+import { OTLPLogExporter } from '@opentelemetry/exporter-logs-otlp-grpc';
+import { Resource } from '@opentelemetry/resources';
+import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
+import { logs } from '@opentelemetry/api-logs';
+import { registerInstrumentations } from '@opentelemetry/instrumentation';
+import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express';
+import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
+import formatter from './formatter';
+import readJsonFile from '../../readJsonFile';
+
+const { buildVersion: version } = readJsonFile('../../../.build-meta.json');
+
+const logMethods = ['error', 'warn', 'log', 'info', 'debug'];
+
+let batchLogProcessor;
+
+export const shutdownOtelLogger = () => {
+  if (process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) return batchLogProcessor.shutdown();
+  return undefined;
+};
+
+const setupTracer = () => {
+  const provider = new NodeTracerProvider();
+  provider.register();
+  registerInstrumentations({
+    instrumentations: [
+      new HttpInstrumentation(),
+      new ExpressInstrumentation(),
+    ],
+  });
+};
+
+export const createOtelLogger = () => {
+  const customResourceAttributes = process.env.OTEL_RESOURCE_ATTRIBUTES ? process.env.OTEL_RESOURCE_ATTRIBUTES.split(';').reduce((acc, curr) => {
+    const [key, value] = curr.split('=');
+    return {
+      ...acc,
+      [key]: value,
+    };
+  }, {}) : {};
+
+  const resource = new Resource({
+    [SemanticResourceAttributes.SERVICE_NAME]: process.env.OTEL_SERVICE_NAME,
+    [SemanticResourceAttributes.SERVICE_NAMESPACE]: process.env.OTEL_SERVICE_NAMESPACE,
+    [SemanticResourceAttributes.SERVICE_INSTANCE_ID]: os.hostname(),
+    [SemanticResourceAttributes.SERVICE_VERSION]: version,
+    ...customResourceAttributes,
+  });
+
+  const collectorOptions = {
+    headers: {},
+    concurrencyLimit: 10,
+    resource,
+  };
+
+  const loggerProvider = new LoggerProvider({ resource });
+  const logExporter = new OTLPLogExporter(collectorOptions);
+
+  if (process.env.NODE_ENV === 'development') {
+    loggerProvider.addLogRecordProcessor(
+      new SimpleLogRecordProcessor(new ConsoleLogRecordExporter())
+    );
+  }
+
+  batchLogProcessor = new BatchLogRecordProcessor(logExporter);
+
+  loggerProvider.addLogRecordProcessor(batchLogProcessor);
+  logs.setGlobalLoggerProvider(loggerProvider);
+
+  const otelLogger = logs.getLogger(process.env.OTEL_SERVICE_NAME);
+
+  // Add trace IDs to logs
+  setupTracer();
+
+  const logger = logMethods.reduce((acc, curr) => {
+    function emitLog(...args) {
+      const logRecord = formatter(curr, ...args);
```

Comment:
logging is usually a hot path, could we test the difference of
```suggestion
    const formatForLevel = formatter.bind(acc, curr);
    function emitLog() {
      const logRecord = formatForLevel.apply(arguments);
```
and/or
```suggestion
    function emitLog() {
      Array.prototype.unshift.call(arguments, curr);
      const logRecord = formatter.apply(acc, arguments);
```

likely not the biggest use of CPU in this call chain though so maybe a heat stack to see if these would be negligible for the cost of current readability \m/

Reply:
[Rest seems more performant than arguments](https://www.measurethat.net/Benchmarks/Show/10518/0/rest-parameters-vs-arguments), and [spread is very close to or marginally better than apply](https://www.measurethat.net/Benchmarks/Show/4383/0/spread-operator-vs-apply)

Reply:
apply with bind doesn't work the way I thought it would, and running simple tests locally (via cronometro, from https://github.com/nodejs/undici/blob/HEAD/benchmarks/benchmark.js#L272) I'm seeing spread to have 2x as many operations as the unshift & apply
Cheers!